Add Seed for policies, policy sets, policy contents and conditions

moolitayer opened the pull request:
Based on read_only behavior added in #8018 

**Seed**
conditions.yml
 - :name: if container image has high severity openscap results  
   :description: if container image has openscap results where result=fail And severity=High

miq_policies.yml
 - :name: analyse incoming container images  
   :description: Analyse incoming container images
 - :name: handle container images with high severity openscap results 
   :description: handle container images with high severity OpenSCAP results

miq_policy_content.yml
 - :miq_policy: analyse incoming container images  
   :miq_action: container_image_analyze  
   :miq_event_definition: containerimage_created
 - :miq_policy: handle container images with high severity openscap results
   :miq_action: compliance_failed  
   :miq_event_definition: containerimage_scan_complete

miq_policy_sets.yml
 - :name: container providers default  
   :description: container providers default

moolitayer commented:
@miq-bot add_label providers/containers, control

moolitayer commented:
@chrisarcand @durandom @Fryguy as suggested in #7413 please reivew
fyi @simon3z

Fryguy commented on db/fixtures/conditions.yml:
```diff
@@ -0,0 +1,24 @@
+---
+- :name: no_high_severity_failures
+  :description: Does not have cve failures with high severity
```
minor, but CVE should be all caps as it's an acronym.

Fryguy replied:
Actually, this condition description is very vague.  It implies we can check for every CVE known even into the future, and that's not accurate.  Since it's really only the CVEs known by OpenSCAP, perhaps "OpenSCAP" should be in the name and description.

Fryguy commented on db/fixtures/miq_policies.yml:
```diff
@@ -0,0 +1,29 @@
+---
+- :name: analyze_new_images
+  :description: Analyze all new images
```
This description and name might want to say "container" somewhere in it, so it's clear it only pertains to container images and not, say, cloud images.

Fryguy replied:
Based on my reasoning [here](https://github.com/ManageIQ/manageiq/pull/7966/files#r59782489), this should probably be OpenSCAP specific in the description and name as well.  Mabye something like "Analyse all new containers images for known vulnerabilities with OpenSCAP ruleset"

Fryguy commented on lib/evm_database.rb:
```diff
@@ -16,6 +16,12 @@ class EvmDatabase
+    Condition
+    MiqPolicy
+    MiqPolicyContent
+    MiqPolicySet
```
Why are these primordial?  primordial are only for things that the system can't even boot without (and yes, by that sentiment I agree that RssFeed and VmdbDatabase should not be in there :smile:)

FYI, Non-primoridal will still get seeded, even if not in this list

moolitayer replied:
This also creates order, although I understand it was not meant for it.

Should I add a ORDERED_CLASSES after PRIMORDIAL_CLASSES and use it [here](https://github.com/ManageIQ/manageiq/blob/master/lib/evm_database.rb#L50)
in a different pr?

Fryguy replied:
Oh I see...hmmm...yeah an ORDERED_CLASSES seems ok.  Another thing that we do elsewhere ,when ordering is important, is to put a seed call to the dependent methods right in the .seed methods itself.  In other words, if `Foo.seed` depends on `Bar.seed` we would write it as follows:

```ruby
class Foo
  def self.seed
    Bar.seed
    # Seeding code for Foo
  end
end

class Bar
  def self.seed
    # Seeding code for Bar
  end
end
```

Since seeding should be reentrant, this should work fine (although it may not be the most effiecient thing depending on how the various seeding code is written.

moolitayer replied:
Submitted ORDERED_CLASSES #8013.

Another way I saw before in some products would be
```ruby
class Foo
  def self.seed
    # Seeding code for Foo
  end

  def self.seed_dependencies
    [Bar]
  end
end
```
Then it is very easy to do topological sort and detect circles: at each stage if there is a class whose all dependencies are met seed it and remove it from dependency lists. If there is not there is surely a circle.

Fryguy replied:
#8013 is merged, so this comment should have gone away as it was originally on PRIMORDIAL_CLASSES...the GitHub diff reorganizer relocated it to here, but this code is fine.

Fryguy commented on spec/models/condition_spec.rb:
```diff
@@ -1,4 +1,24 @@
+
+      expect(Condition.count).to eq(1)
+      Condition.all.each do |condition|
+        spec = specifications.pop
```
Even though there's only 1 here, if this were changed to more than one, then this test would be non-deterministic because of the unknown database ordering.  In the case of conditions, the name is the unique bit, so you could probably change this to something like the following

```ruby
specifications = YAML.load_file(File.join(ApplicationRecord::FIXTURE_DIR, "#{Condition.table_name}.yml"))
specifications_by_name = specifications.index_by { |h| h[:name] }

expect(Condition.count).to eq(1)
Condition.all.each do |condition|
  spec = specifications_by_name[condition.name]
  ...
```

You could also do a sort by name on both sides, but then you have to deal with case and collation and in the end it's probably better to just to the Hash indexing.

Fryguy replied:
Same comment goes for the other specs since it's the same pattern.

Fryguy commented on spec/models/condition_spec.rb:
```diff
@@ -1,4 +1,24 @@
 describe Condition do
+  context ".seed" do
```
Minor, but prefer `describe ".seed" do` over a `context` in this case.  It's not a real official pattern yet, but most of the RSpec tests are moving that way.

Fryguy commented on db/fixtures/miq_policies.yml:
```diff
@@ -0,0 +1,29 @@
+  :mode: control
+  :conditions: []
+- :name: handle_high_severity_failures
+  :description: Prevent images with high severity cves from running at openshift
```
Same here...since it's container specfic it should say "container" somewhere in the name and description.  Additionally, CVE should be caps (probably like "CVEs" in this case), and OpenShift is a product name so it should be properly capitalized as OpenShift.  You *might* have to make it fully spelled out as "Red Hat OpenShift" or "OpenShift Enterprise" unless it's not really Enterprise specific (you get the idea).

Fryguy commented on db/fixtures/miq_policy_sets.yml:
```diff
@@ -0,0 +1,8 @@
+---
+- :name: container_image_default
+  :description: Container Image default profile
+  :guid:
```
You can just leave this out entirely, I think.  We don't want anyone to populate it by accident anyway.

Fryguy replied:
Same goes for the other fixtures as well.

moolitayer commented:
Addressed comments from @Fryguy 
Thanks for the feedback :eyeglasses:

moolitayer commented on app/models/miq_policy_content.rb:
```diff
@@ -3,6 +3,50 @@ class MiqPolicyContent < ApplicationRecord
+      :miq_policy       => policy,
+      :success_sequence => success_sequence,
+      :failure_sequence => failure_sequence
+    )
```
Is this really the primary key of MiqPolicyContent? @Fryguy @jrafanie ?

jrafanie replied:
@moolitayer 😕  I'm not familiar with that model.  It looks to be mostly unchanged since it was added, maybe @gtanzillo knows.

moolitayer commented:
@Fryguy @jrafanie In this pr there are entities the user can change (miq policy sets, miq policies, miq policy contents and conditions) that will magically 'return to default' after seeding. What should I do?
Avoid updating changed records? add a read only field?

blomquisg commented:
> What should I do? Avoid updating changed records? add a 'read only' field + behavior?

I assumed the intent of seeding is that you only do it after setting up the database.  If someone attempts to seed after the database has been in use, then I would think they're asking for problems.

@Fryguy, is that a correct assumption?

If that's the case, then I think we don't really have to do anything.  At best we should add something to the `EvmDatabase.seed` method to indicate that seeding a database that's been "in use" could lead to data loss.

Thoughts?

Fryguy commented:
> If someone attempts to seed after the database has been in use, then I would think they're asking for problems.

No that's not right...seeding should be able to be run over and over to get the default data to a consistent state (e.g. add new records on product update, update the descriptions of other records, etc).

> In this pr there are entities the user can change

Ugh...we have avoided trying to seed these table in the past because that's a pain to deal with :(  I didn't realize these tables fell into that category when I first reviewd...sorry about that. Generally, if a user can make changes, there are few approaches.

1. Make seed data read-only and put the word "Sample" or "Example" in the description.  This way, the user can copy them and then make changes.  If a `read_only` or `system` column is not in place, then it will need to be added, and the UI will have to change accordingly to not allow editing of those entries.
2. Have a column that indicates that the original entry has been modified by the user, and thus is not a candidate for update during seeding.

@h-kataria @gmcculloug How have you dealt with this in the past?  I know this has happened with other models, but I can't think of which specific one I could point @moolitayer to as an example.

h-kataria commented:
@fryguy we seed sample read only Analysis Profiles that user cannot delete or edit, they can be copied in UI to create new ones in this case read_only column is set to true. These are seeded using a constant from model instead of from a yaml file but you can look at seeding in https://github.com/ManageIQ/manageiq/blob/master/app/models/scan_item.rb to get an idea on how that works.
Other lines of interest:
https://github.com/ManageIQ/manageiq/blob/master/app/models/scan_item.rb#L9
https://github.com/ManageIQ/manageiq/blob/master/app/models/scan_item.rb#L58

gmcculloug commented:
@Fryguy In the past we have had a column to mark that the data that has been seeded and the UI treats it as read-only to avoid this issue.  Users can use the data as-is or copy to a new instance and modify.  I am thinking of the provisioning dialogs and as @h-kataria points out the Analysis Profiles, but there are other examples.

I would think we would want the same thing here.

kbrock commented:
The root tenant record is seeded but can be changed. Maybe vmdb too. Some seeds just care that they are present while others want to be identical

kbrock commented:
Seeds need to be idem potent. I wang them to get run after every upgrade.

Heck, some get run on every boot of every worker. (Not a fan, but it is what we do

kbrock commented:
The root tenant record is seeded but can be changed. Maybe vmdb too. Some seeds just care that they are present while others want to be identical. But I guess that data is from ruby not yaml

moolitayer commented:
@Fryguy @blomquisg @h-kataria @gmcculloug @kbrock Thanks for the helpful feedback.
Bearing in mind approaching dedalines I've submitted #8012 with the read_only migration change and #8018 with the behavior change for the control screens. This pr's ymls were updated with the read_only attribute. Please review

moolitayer commented:
@miq-bot add_label wip

moolitayer commented:
fyi @simon3z

simon3z commented:
> fyi @simon3z

Got it @moolitayer thanks! We'll continue on #8012 for now (and @8018 later).

moolitayer commented:
@miq-bot remove_label wip

jrafanie commented on app/models/condition.rb:
```diff
@@ -33,6 +33,25 @@ def self.conditions
+        _log.info("Creating Condition: [#{condition[:name]}] ")
+        Condition.create(condition)
+      else
+        rec.attributes = condition
```
What happens if attributes contains key/values that aren't in condition?

jrafanie commented on app/models/miq_policy.rb:
```diff
@@ -369,6 +370,38 @@ def self.seed
+      _log.info("Creating MiqPolicy: [#{policy[:name]}] ")
+      rec = MiqPolicy.create(policy)
+    else
+      rec.attributes = policy
```
Same as above.  I think it should do the right thing, but maybe we could add a test to ensure the update doesn't clear out existing values that weren't in the hash assigned to attributes.

moolitayer replied:
added the test at [condition_spec.rb](https://github.com/moolitayer/manageiq/blob/e19670e93dd6408cfa461a76ac3a5bc46b7a3a7e/spec/models/condition_spec.rb#L23)

jrafanie commented on app/models/miq_policy_set.rb:
```diff
@@ -116,11 +116,44 @@ def self.import_from_yaml(fd)
+      _log.info("Creating MiqPolicySet: [#{policy_set[:name]}] ")
+      rec = MiqPolicySet.create(policy_set)
+    else
+      rec.attributes = policy_set
```
Same

jrafanie commented on app/models/miq_policy_content.rb:
```diff
@@ -3,6 +3,50 @@ class MiqPolicyContent < ApplicationRecord
+    rec.attributes = content
+    if rec.changed?
+      _log.info("Updating MiqPolicyContent: [#{policy_name}, #{content[:success_sequence]},"\
+                    " #{content[:failure_sequence]}]")
```
How useful is it to include the success_sequence and failure_sequence in the log line?  Maybe the policy_name and rec.id would be enough so we don't need the `\` to escape the CR.

moolitayer commented:
Address comments from @jrafanie

moolitayer commented on db/fixtures/miq_policies.yml:
```diff
@@ -0,0 +1,29 @@
+---
+- :name: Analyse all new container images for known vulnerabilities with OpenSCAP ruleset
```
@jrafanie @Fryguy Can all names be shorter and underscored and the descriptions have the long version? The user only sees the description, the name is used for linking entities (and adding the default policy to new container providers in #7567)

jrafanie replied:
@moolitayer typically names are not underscored, why can't we just a short name and have a long description?  What does the underscore buy us?

moolitayer replied:
> @moolitayer typically names are not underscored, why can't we just a short name and have a long description? What does the underscore buy us?

Nothing really, updated names, thanks.

moolitayer commented:
@jrafanie @Fryguy another round please.
Thank's for the feedback so far!

moolitayer commented:
@jrafanie @Fryguy Changed event names & descriptions and added to pr message.
Changed logic of miq_policy_content seed.

moolitayer commented:
ping @chessbyte @blomquisg @simon3z

kbrock commented:
are these fixtures sample data? Can a customer use the appliance without these?
Also, is it legit for a customer to delete any of these fixtures?

If either of those are true, lets please evaluate our best strategy here.

moolitayer commented:
> are these fixtures sample data? Can a customer use the appliance without these?
> Also, is it legit for a customer to delete any of these fixtures?

They are not sample data, but default policies. This pr is based on read_only attribute added for policy entities in #8018. A user can not delete/modify these policies through the UI but he should still be able to use the appliance normally in case they are somehow gone (Again, should not happen)

jrafanie commented on app/models/miq_policy.rb:
```diff
@@ -369,6 +370,38 @@ def self.seed
+    rec
+  end
+
+  def self.seed_conditions(rec, conditions)
```
For clarity, maybe this should be called `seed_link_to_conditions` or a clearer name (naming is hard)?  We're assuming that conditions are already seeded by this point, the current method name confused me at first.

jrafanie replied:
rec is a MiqPolicy, maybe rename it to make that clear?

jrafanie commented on app/models/miq_policy.rb:
```diff
@@ -369,6 +370,38 @@ def self.seed
+    end
+  end
+
+  def self.seed_attributes(policy)
```
We're really seeding_policies or seed_policies_from_attributes here right?  Maybe also change the parameter to policy_attributes or policy_fixture or something so we know it's not a MiqPolicy object.

jrafanie commented on app/models/miq_policy_content.rb:
```diff
@@ -3,6 +3,53 @@ class MiqPolicyContent < ApplicationRecord
 
+  def self.seed
+    fixture_file = File.join(FIXTURE_DIR, "miq_policy_contents.yml")
+    policy_contents = File.exist?(fixture_file) ? YAML.load_file(fixture_file) : []
```
Maybe I'm strange but I read policy_contents as MiqPolicyContent objects, not as fixtures/attributes used to create those objects.  Maybe content_attributes or content_fixtures?

jrafanie commented on app/models/miq_policy_content.rb:
```diff
@@ -3,6 +3,53 @@ class MiqPolicyContent < ApplicationRecord
+    policy_contents = File.exist?(fixture_file) ? YAML.load_file(fixture_file) : []
+
+    policy_contents.each do |content|
+      find_associations!(content)
```
base on the above comment, content would be something like content_attributes or something to make it clear it's the attributes we're using to find/create/update a MiqPolicyContent.

jrafanie replied:
I'd prefer this not modify the receiver so it's easier to test and not have side effects.  Maybe merge the result of this method.

jrafanie commented on app/models/miq_policy_content.rb:
```diff
@@ -3,6 +3,53 @@ class MiqPolicyContent < ApplicationRecord
+      find_associations!(content)
+      policy_name = content[:miq_policy].try(:name)
+
+      rec = find_by_pkey(policy_name, content[:success_sequence], content[:failure_sequence])
```
I don't know if this is a primary key but I agree, I'm cautious about this one.

jrafanie replied:
@gtanzillo Is this the "way" to lookup MiqPolicyContent for seeding?

jrafanie commented on app/models/miq_policy_content.rb:
```diff
@@ -3,6 +3,53 @@ class MiqPolicyContent < ApplicationRecord
+      if rec.nil?
+        create_seed_record(policy_name, content)
+      else
+        update_seed_record(rec, policy_name, content)
```
I probably wouldn't create methods for the create/update work.  If I were to extract a method, I would create a method to seed one content...

```ruby
...
policy_contents.each do |content|
  seed_one_policy_content(content)
end
```

moolitayer replied:
I started with a method to seed one policy content but I had to split it further to find_by_pkey, create_seed_record and update_seed_record to get it simple enough for rubocop's taste

jrafanie commented on app/models/miq_policy_set.rb:
```diff
@@ -116,11 +116,44 @@ def self.import_from_yaml(fd)
+    fixtures = File.exist?(fixture_file) ? YAML.load_file(fixture_file) : []
+
+    fixtures.each do |policy_set|
+      rec = seed_attributes(policy_set.except(:miq_policies))
```
Same, seed_policy_sets or something

jrafanie commented on app/models/miq_policy_set.rb:
```diff
@@ -116,11 +116,44 @@ def self.import_from_yaml(fd)
+
+    fixtures.each do |policy_set|
+      rec = seed_attributes(policy_set.except(:miq_policies))
+      seed_policies(rec, policy_set[:miq_policies])
```
same, seed_link_to_policy or something

jrafanie commented:
LGTM @moolitayer, I had mostly "naming" questions/comments, feel free to disagree or offer better suggestions.  The pkey lookup method feels weird as you suggest, how do we find/locate the policy_content elsewhere?

moolitayer commented:
> The pkey lookup method feels weird as you suggest, how do we find/locate the policy_content elsewhere?

@jrafanie I could not find anywhere relevant. Addressed comments, thanks for the feedback!

jrafanie commented on app/models/miq_policy.rb:
```diff
@@ -369,6 +370,38 @@ def self.seed
+    fixtures = File.exist?(fixture_file) ? YAML.load_file(fixture_file) : []
+
+    fixtures.each do |policy_fixture|
+      rec = seed_policies(policy_fixture.except(:conditions))
```
Should this be singular since we're only seeing one policy at a time?  Sorry.  😢

jrafanie commented on app/models/miq_policy_set.rb:
```diff
@@ -116,11 +116,44 @@ def self.import_from_yaml(fd)
+    fixtures = File.exist?(fixture_file) ? YAML.load_file(fixture_file) : []
+
+    fixtures.each do |policy_set_fixture|
+      rec = seed_policy_sets(policy_set_fixture.except(:miq_policies))
```
Same as ☝️, singular... since the method only creates/finds a single policy set.

jrafanie commented on spec/models/miq_policy_content_spec.rb:
```diff
@@ -0,0 +1,23 @@
+      MiqPolicyContent.all.each do |mpc|
+        spec = specifications_by_pkey["#{mpc.miq_policy.name}_#{mpc.success_sequence}_#{mpc.failure_sequence}"]
+        expect(mpc).to have_attributes(
+          spec.except(:created_on, :updated_on, :miq_policy, :miq_action, :miq_event_definition)
```
I don't know if this list will grow, but it might be a pain to maintain,  keep it in mind if we have to alter this list of columns.

I think we could test one or two attributes we care about instead of checking all attributes not in the this exclude list.

jrafanie replied:
In other words, I don't know what we gain from testing all of these attributes over just checking `:name` or whatever if useful.  We don't really have to test attributes= and save except for things we're concerned about (missing attributes aren't updated, unchanged attributes aren't saved, etc.)

jrafanie commented:
@moolitayer Please rebase (to fix the test failure) and comment if you want to make any more changes based on my comments.  I think it's ready for merge after it's green.

moolitayer commented:
Addressed comments, Thanks @jrafanie

moolitayer commented:
@miq-bot add_label darga/yes

moolitayer commented:
Changed seed_link_to_conditions as I noticed it wasn't idempotent;
on each seed call it added a link to the same conditions again and again

moolitayer commented:
@jrafanie can this merge with the travis migration failure or do you prefer to wait for #8291

chessbyte commented:
Now that #8291 is merged, closed/reopened to restart Travis

Fryguy commented:
@gmcculloug Since this is in control, can you do a final review before merge?  I think you might have had higher level concerns about using policies at all for driving workflows like this, but perhaps I'm mistaken?

moolitayer commented:
@gmcculloug I'm not aware of any concerns raised at any point regarding this patch (as appose to #8097) could you please respond?

gmcculloug commented:
@moolitayer I realize it is way late to be ask this now, but since we support import/export of policy sets (and related components) into a single file is there a reason we are not using that for the basis of seeding policies?  Was this considered and maybe not viable for some reason?

moolitayer commented:
@gmcculloug a defined objective was that the OpenSCAP integration would work out of the box. It was discussed in a meeting and the conclusion was to seed Policy data. 

Now I see that Adding another mechanism is causing concern and I completely understand that concern.

At first step I suggest replacing darga/yes with darga/no and we can consider this in the future if the need arises.

I'm in the process of documenting suggested policies and import instructions. Just like @jonnyfiveiq has done in the past for Shellshock (or was it Heartbleed?)
currently in https://github.com/ManageIQ/manageiq/issues/7690#openscap_profile

gmcculloug commented:
@moolitayer My question was really asking why you implemented new methods for importing this type of data instead of using the existing import method (from a new seed method) which also has the advantage of capturing all the policy configuration in a single file.  Seems there is a chance for a lot of code re-use here instead of supporting two styles of imports.

moolitayer commented:
oh I'm sorry for misunderstanding you there @gmcculloug.

The point I raised at the time toward the existing representation was that since it is nested It is not ideal for seed as it cannot express:
- nested entities that do not belong to a parent, for example a condition that does not belong to a policy (i.e seeded as an example condition)
- nested entities that belong to more then one entity (policy in several profiles, condition in several policies) 

Now that you raised this I found out I was wrong regarding the second item. I gave this another go. There were some other concerns raised by @chessbyte and @Fryguy that need attention
- do we see examples or assignable policies? - I see assignable policies
- are they read only? - currently no since we do not intend attaching them by default
- note I had to seed some guids with this approach since AFAIK [existing code](https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_policy/import_export.rb#L67) needs those

@gmcculloug @chessbyte @Fryguy please review

gmcculloug commented on lib/evm_database.rb:
```diff
@@ -20,6 +20,9 @@ class EvmDatabase
     MiqWidget
+    MiqAction
+    MiqEventDefinition
+    MiqPolicySet
```
@moolitayer Do you require any of these classes defined in this list?  They will be dynamically called from here: [lib/evm_database.rb#L47](https://github.com/ManageIQ/manageiq/blob/6f8773c8205d0d36d5ca914a51c92e439987f59d/lib/evm_database.rb#L47)

moolitayer replied:
The policy seeding relays on MiqAction and MiqEventDefinition being seeded before. 
Now that this patch only has MiqPolicySet.seed I can drop this since it is lexicographically last or should I keep it for explicitness @gmcculloug ?

gmcculloug replied:
It is better to keep them in the `ORDERED_CLASSES` list.  Thanks.

moolitayer commented:
Changed seeding to be read_only to give #7567 a chance

moolitayer commented:
@Fryguy @chessbyte I'd like to know if this approach is more agreeable given past objections

gmcculloug commented:
Testing this a bit in the UI after seeding the UI is properly honoring the read-only flag and does not allow you to delete the item.  The one thing I did notice is if you copy the policy it retains the read-only flag and the new item is still read-only.  We can address this in a separate PR.

@chessbyte @Fryguy I am good with merging this if you are.

chessbyte commented:
@gmcculloug if it no longer auto-assigns the policy, I am 👍 with having read-only samples seeded.
